mne/simulation/metrics.py

```diff
@@ -63,3 +67,310 @@ def source_estimate_quantification(stc1, stc2, metric='rms'):
+        vert = np.union1d(vert1, vert2)
+        data1 = np.zeros([len(vert), stc1.data.shape[1]])
+        data2 = np.zeros([len(vert), stc2.data.shape[1]])
+        data1[np.searchsorted(vert, vert1)] = \
+            stc1.data[idx_start1:idx_start1 + len(vert1)]
+        data2[np.searchsorted(vert, vert2)] = \
+            stc2.data[idx_start2:idx_start2 + len(vert2)]
+        idx_start1 += len(vert1)
+        idx_start2 += len(vert2)
+        stc1.vertices[i] = vert
+        stc2.vertices[i] = vert
+        all_data1.append(data1)
+        all_data2.append(data2)
+
+    stc1._data = np.concatenate(all_data1, axis=0)
+    stc2._data = np.concatenate(all_data2, axis=0)
+    return stc1, stc2
+
+
+def _apply(func, stc_true, stc_est, per_sample):
+    """Apply metric to stcs.
+
+    Applies a metric to each pair of columns of P and Q
+    if per_sample is True. Otherwise it applies it to P and Q
+    directly.
+    """
+    if per_sample:
+        metric = np.zeros(stc_true.data.shape[1])  # one value per time point
+        for i in range(stc_true.data.shape[1]):
+            metric[i] = func(stc_true.data[:, i:i + 1],
+                             stc_est.data[:, i:i + 1])
+    else:
+        metric = func(stc_true.data, stc_est.data)
+    return metric
+
+
+def _thresholding(stc_true, stc_est, threshold):
+    if isinstance(threshold, str):
+        t = _check_threshold(threshold)
+        stc_true._data[np.abs(stc_true._data) <=
+                       t * np.max(np.abs(stc_true._data))] = 0.
+        stc_est._data[np.abs(stc_est._data) <=
+                      t * np.max(np.abs(stc_est._data))] = 0.
+    else:
+        stc_true._data[np.abs(stc_true._data) <= threshold] = 0.
+        stc_est._data[np.abs(stc_est._data) <= threshold] = 0.
+    return stc_true, stc_est
+
+
+def _cosine(x, y):
+    p = np.reshape(x, (-1, 1))
+    q = np.reshape(y, (-1, 1))
+    return np.dot(p.T, q) / (norm(p) * norm(q))
+
+
+@fill_doc
+def stc_cosine_score(stc_true, stc_est, per_sample=True):
+    """Compute cosine similarity between 2 source estimates.
+
+    Parameters
+    ----------
+    %(metric_stc_true)s
+    %(metric_stc_est)s
+    %(metric_per_sample)s
+
+    Returns
+    -------
+    %(stc_metric)s
+    """
+    stc_true, stc_est = _uniform_stc(stc_true, stc_est)
+    func = partial(_cosine)
+    metric = _apply(func, stc_true, stc_est, per_sample=per_sample)
+    return metric
+
+
+def _check_threshold(threshold):
+    """Accept a float or a string that ends with %."""
+    if isinstance(threshold, str):
+        if threshold.endswith("%"):
+            return float(threshold[:-1]) / 100.0
+        else:
+            raise ValueError('Threshold if a string must end with '
+                             '"%%". Got %s.' % threshold)
+    else:
+        return threshold
+
+
+def _dle(p, q, src, stc):
+    """Aux function to compute dipole localization error."""
+    from sklearn.metrics import pairwise_distances
+    p = np.sum(np.abs(p), axis=1)
```

Comment:
Perhaps define a threshold parameter, as this formulation can be too conservative for non-sparse estimates ?

Reply:
we won't find one metric that fit all needs, hence the need for implementing all these. We should implement what are the metrics commonly used in the literature.

I am more worried about our handling of the sign or the temporal dimension.

Reply:
here are few refs:

https://www.ncbi.nlm.nih.gov/pubmed/16271483
https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3018574/

cc @olafhauk